fix: Hot reload always crashes on Static Mode after a syntax error is introduced at one point, even when it's fixed afterwards

**Description**

When running development (`jaspr serve`) on **Static Mode**, hot reload always crashes and can't recover after at one point there's a syntax error in the code (even after the syntax error is fixed afterwards).

**Steps To Reproduce**

1. Create a new project using `jaspr create my_website`.
2. Choose `static` rendering mode (which should be the default).
3. Let other setup options with their default selections.
4. Run `jaspr serve`.
5. Open the page in a browser.
6. Introduce a syntax error. For example, you can add a component to the middle of a list without the mandatory separator comma:
    ```dart
    yield section([
      img(src: 'images/logo.png', width: 80),
      h1([text('Welcome')]),
      // Syntax error below (missing comma)
      Counter()
      p([text('You successfully create a new Jaspr site.')]),
      div(styles: Styles.box(height: 100.px), []),
      const Counter(),
    ]);
    ```
8. Notice that hot reload crashes, which is **expected**.
    ```
    [BUILDER] [ERROR] This builder requires Dart inputs without syntax errors.
    [BUILDER] [ERROR] However, package:jas_static/pages/home.dart (or an existing part) contains the following errors.
    [BUILDER] [ERROR] home.dart:16:7: Expected to find ','.
    [BUILDER] [ERROR]
    [BUILDER] [ERROR] Try fixing the errors and re-running the build.
    ```
10. Fix the syntax error (add comma after `Counter()`).
11. Notice that hot reload **still crashes**, even when there are no more syntax errors.
    ```
    [SERVER] [ERROR] ERROR - 2024-04-08 13:11:15.623075
    [SERVER] [ERROR] GET /
    [SERVER] [ERROR] Error thrown by handler.
    [SERVER] [ERROR] ClientException: HTTP request failed. Client is already closed., uri=http://localhost:5567/
    [SERVER] [ERROR] package:http/src/io_client.dart 110:7            IOClient.send
    [SERVER] [ERROR] package:shelf_proxy/shelf_proxy.dart 59:48       proxyHandler.<fn>
    [SERVER] [ERROR] package:shelf_gzip/src/gzip_encoding.dart 51:44  createGzipMiddleware.<fn>.<fn>.<fn>
    [SERVER] [ERROR] dart:async                                       new Future.sync
    [SERVER] [ERROR] package:shelf_gzip/src/gzip_encoding.dart 51:21  createGzipMiddleware.<fn>.<fn>
    [SERVER] [ERROR] ERROR - 2024-04-08 13:11:15.644062
    [SERVER] [ERROR] GET /favicon.ico
    ```
12. If you refresh the page on the browser, it will show you "Internal Server Error".

**Doctor Output**

```
[✓] Jaspr CLI (Version 0.11.1)
  • Dart Version 3.3.1 (stable) (Wed Mar 6 13:09:19 2024 +0000) on "windows_x64" at dart
  • Running on windows "Windows 10 Home Single Language" 10.0 (Build 22631) - Locale en-ID
  • Analytics: Enabled

[✓] Current Project
  • Dependencies on core packages:
    • jaspr: ^0.11.1
    • jaspr_builder: ^0.11.1 (dev)
    • jaspr_web_compilers: ^4.0.9 (dev)
    • jaspr_router: ^0.4.0
  • Rendering mode: static
  • Uses jaspr compilers: true
  • Uses flutter embedding: false

```

**Expected Behavior**

Hot reload should work normally after the syntax error is fixed.

**Additional Context**

This error does not happen when the project is running on **Server Mode**.

--- a/packages/jaspr/lib/src/server/server_app.dart
+++ b/packages/jaspr/lib/src/server/server_app.dart
@@ -22,14 +22,27 @@ class ServerApp {
 
   final SetupFunction _setup;
 
+  static Object? _runLock;
   static HttpServer? _server;
   static http.Client? _client;
 
   void _run() async {
     var isFirstStartup = _server == null;
 
-    await _server?.close(force: true);
-    _server = await _createServer();
+    var lock = _runLock = Object();
+    var (client, server) = await _createServer();
+
+    if (_runLock != lock) {
+      server.close(force: true);
+      client.close();
+      return;
+    }
+
+    _server?.close(force: true);
+    _server = server;
+
+    _client?.close();
+    _client = client;
 
     if (isFirstStartup) {
       _writePid();
@@ -45,12 +58,11 @@ class ServerApp {
     }
   }
 
-  Future<HttpServer> _createServer() {
+  Future<(http.Client, HttpServer)> _createServer() async {
     var port = int.parse(Platform.environment['PORT'] ?? '8080');
-    _client?.close();
-    _client = http.Client();
-    var handler = createHandler((_, render) => render(_setup), client: _client);
-    return shelf_io.serve(handler, InternetAddress.anyIPv4, port, shared: true);
+    var client = http.Client();
+    var handler = createHandler((_, render) => render(_setup), client: client);
+    return (client, await shelf_io.serve(handler, InternetAddress.anyIPv4, port, shared: true));
   }
 
   void _writePid() {

--- a/packages/jaspr/lib/src/server/server_handler.dart
+++ b/packages/jaspr/lib/src/server/server_handler.dart
@@ -130,11 +130,11 @@ Handler _sseProxyHandler(http.Client client, String webPort) {
   }
 
   return (Request req) async {
-    if (req.headers['accept'] == 'text/event-stream' && req.method == 'GET') {
-      return createSseConnection(req);
-    }
+      if (req.headers['accept'] == 'text/event-stream' && req.method == 'GET') {
+        return await createSseConnection(req);
+      }
 
-    return Response.notFound('');
+      return Response.notFound('');
   };
 }
 

--- a/packages/jaspr/lib/src/server/server_handler.dart
+++ b/packages/jaspr/lib/src/server/server_handler.dart
@@ -130,11 +130,11 @@ Handler _sseProxyHandler(http.Client client, String webPort) {
   }
 
   return (Request req) async {
-      if (req.headers['accept'] == 'text/event-stream' && req.method == 'GET') {
-        return await createSseConnection(req);
-      }
+    if (req.headers['accept'] == 'text/event-stream' && req.method == 'GET') {
+      return await createSseConnection(req);
+    }
 
-      return Response.notFound('');
+    return Response.notFound('');
   };
 }
 

--- a/packages/jaspr_builder/lib/src/client/client_module_builder.dart
+++ b/packages/jaspr_builder/lib/src/client/client_module_builder.dart
@@ -16,6 +16,8 @@ class ClientModuleBuilder implements Builder {
   FutureOr<void> build(BuildStep buildStep) async {
     try {
       await generateClientModule(buildStep);
+    } on SyntaxErrorInAssetException catch (e) {
+      rethrow;
     } catch (e, st) {
       print('An unexpected error occurred.\n'
           'This is probably a bug in jaspr_builder.\n'

--- a/packages/jaspr_builder/lib/src/client/client_module_builder.dart
+++ b/packages/jaspr_builder/lib/src/client/client_module_builder.dart
@@ -16,7 +16,7 @@ class ClientModuleBuilder implements Builder {
   FutureOr<void> build(BuildStep buildStep) async {
     try {
       await generateClientModule(buildStep);
-    } on SyntaxErrorInAssetException catch (e) {
+    } on SyntaxErrorInAssetException {
       rethrow;
     } catch (e, st) {
       print('An unexpected error occurred.\n'

--- a/packages/jaspr_builder/pubspec.yaml
+++ b/packages/jaspr_builder/pubspec.yaml
@@ -14,7 +14,7 @@ topics:
   - build-runner
 
 dependencies:
-  analyzer: ^5.4.0
+  analyzer: '>=5.12.0 <7.0.0'
   build: ^2.0.0
   collection: ^1.15.0
   dart_style: ^2.2.3

--- a/packages/jaspr_cli/lib/src/helpers/proxy_helper.dart
+++ b/packages/jaspr_cli/lib/src/helpers/proxy_helper.dart
@@ -115,7 +115,7 @@ Handler _sseProxyHandler(http.Client client, String webPort, Logger logger) {
 
   return (Request req) async {
     if (req.headers['accept'] == 'text/event-stream' && req.method == 'GET') {
-      return createSseConnection(req);
+      return await createSseConnection(req);
     }
 
     return Response.notFound('');

